[decision] internal cache (#4641)
ElektraInitiative/libelektra: doc/decisions/1_in_discussion/internal_cache.md

```

It was found unexpected that this assert will fail.

### Fewer Keys

When doing a second `kdbGet` with a new keyset no keys will be returned when no backends report changed data, because kdb internally thinks the data is already up-to-date.
A unit test by @atmaxinger:

```c
static void test_doubleGet (void)
{
	printf("running %s\n", __func__);

	// Setup
	Key * parentKey = keyNew("/somewhere", KS_END);
	KeySet * ks = ksNew(0, KS_END);
	KDB * kdb = kdbOpen (ksNew(0, KS_END), parentKey);
	kdbGet (kdb, ks, parentKey);
	ksAppendKey (ks, keyNew ("user:/somewhere", KEY_VALUE, "abc", KEY_END));
	ksAppendKey (ks, keyNew ("user:/somewhere/key", KEY_VALUE, "xyz", KEY_END));
	kdbSet (kdb, ks, parentKey);
	kdbClose (kdb, parentKey);

	// Scenario
	kdb = kdbOpen (ksNew(0, KS_END), parentKey);

	KeySet * ks1 = ksNew (0, KS_END);
	KeySet * ks2 = ksNew (0, KS_END);

	kdbGet (kdb, ks1, keyNew("/somewhere", KEY_END));
	succeed_if (ksLookupByName (ks1, "/somewhere/key", 0) != NULL, "should find key (1)");
	kdbGet (kdb, ks2, keyNew("/somewhere", KEY_END));
	succeed_if (ksLookupByName (ks2, "/somewhere/key", 0) != NULL, "should find key (2)");
}
```

> > It actually outputs should find key (2) so the assertion fails.

Reviewer: This is also said in the next sentence.

Author: [0734b46](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/0734b46b19d14ec65636f3428c65435d6f2bb06f)

It was found unexpected that the second assert will fail.

## Constraints

- memory consumption must be low for `kdbGet`, see [4. Goal: Performance](/doc/GOALS.md), in particular, deep duplication is too expensive

## Assumptions

## Considered Alternatives

### Keep Current Situation

Improve documentation to make people more aware of these two problems:

- add a tutorial about `kdbGet` semantics
- add full examples how to correctly work with `kdbGet`

### Cachefilter Plugin

Naively one would simply cache the whole keyset and use `ksBelow` to always get the keyset.

> This idea was implemented and later on discarded: a3d95f07160d792fdd0f169d8543138c32a2f580

Reviewer: Please add a link here.

Author: [0ca9f78](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/0ca9f785b7d7ce8c70c8c81bebedcd8e3f20675f)

The main problems are:

- very high memory consumption (duplication of KeySets)
- problems specific to [hooks](../4_partially_implemented/hooks.md), see [#1072](https://issues.libelektra.org/1072)

### MMAP Cache with parent key

We make the mmap cache non-optional so that we always have a keyset of configuration data internally.
From this keyset, we use `ksBelow` to return the correct keyset.

**Cons:**

- invalidation of OPMPHM

### MMAP Cache without parent key

We make the mmap cache non-optional and only use a single cache, caching everything.
We remove the parent key of `kdbGet` and `kdbSet` and always return the keyset of the whole KDB.

> 

Reviewer: ```suggestion
Advantage: then also symlinks and constraints for keys outside of the parentKey would work.

```

Reviewer 2: Would that really be a good thing? Reach up out of the mountpoint breaks isolation in any case. It can also only work, within hook plugins, because mountpoints are still isolated for mounted plugins.

Reviewer: symlinks are not affected by the problem you describe. I updated the suggestions to make this more clear. I agree that this is actually a separate problem but it would be solved by some mmap solutions proposed here.

Reviewer 2: What exactly do you mean by "symlinks"?

`kdb get -a` is exactly the issue I describe. I normally expect mountpoints to be isolated. Having a key `system:/foo/bar` refer to `../../abc` (in whatever way) when `system:/foo` is a mountpoint is always fishy. The other way around e.g. `system:/abc` referring to `system:/foo/bar` (assuming `system:/abc` is _not_ a mountpoint) would be less fishy, but still a bit weird.

Author: Is there a resolution to this?

Reviewer: I updated the suggestion already before, it explains symlinks now.

The other stuff is imho another decision and not really important as symlinks already show the whole problem.

Reviewer 2: You still haven't addressed why this would be actually be a good default behaviour. You just claim it is an advantage without any explanation.

Like I said, I would normally expect mountpoints to be isolated. If symlinks work like this, the isolation is partially broken. One could argue that the problem right now is that such "broken" symlinks are not prevented. But your solution doesn't completely fix the problem either.

Take a variant of my previous example:
- `system:/foo` is a mountpoint, there are no other mountpoints
- `spec:/foo/bar` refers to `system:/abc` via `meta:/override`

Now in a plugin that is part of the mountpoint `system:/foo` doing a `ksLookupByName (ks, "/foo/bar", 0)` would **NOT** use `system:/abc`, because that key is **never** part of the keyset passed to this plugin.

However, in an application that used `system:/foo` as the parent, the `meta:/override` **would work** with your proposal. To me that seems like very confusing behaviour, because both the application and the plugin seemingly use the same parent key.

Reviewer: This is a separate decision. Can you please create one?

Author: I added this discussion to the decision.

  However, we need to document for the users of Elektra that keys returned from `kdbGet` are only valid until `kdbClose`.
  If they want to continue using them afterwards, they'd have to deep copy them.
  
  Triggering the delete problem:
  ```c
  Key * originalKey;
  Key * copiedKey;
  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
  
  assert (keyString (copiedKey) == keyString (originalKey));
  assert (keyName (copiedKey) == keyName (originalKey));

  keyDel (originalKey);
  
  keyString (copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to freed memory
  keyName (copiedKey);   // Error! Original name has been deleted.
  ```
 
  Triggering the update problem:
  ```c
  Key * originalKey;
  Key * copiedKey;
  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
  
  assert (keyString(copiedKey) == keyString(originalKey));
  keySetString (originalKey, "new value!");
  
  keyString(copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to potentially freed memory
  ```

>   The same problems in principle exist for `mmapstorage`, however this is just a storage plugin and the user does not have any API to create COW keys themselves.

Reviewer: Not only principle, I think @kodebach is right that `kdbClose` will unmap everything, so the KeySet afterwards is broken.

I think this is currently also not properly documented, so we definitely must do something about it.

```suggestion
  The same problems exist for `mmapstorage` where `kdbClose` frees (munmap) the KeySet.
```

Reviewer 2: As stated elsewhere, I checked and `munmap` actually happens in `kdbSet` for `mmapstorage`. Note: I have not checked when this would be called for the cache.

Author: [217dff4](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/217dff470ccf40b28993628aaf6d91eabdf2d545)

Reviewer: @mpranj can you help out here?

(That `kdbSet` does it might be true but is totally uninteresting for the API user.)

Reviewer 2: > That `kdbSet` does it might be true but is totally uninteresting for the API user

It's not phrased perfectly. But it should tell the API user, that the keyset is only valid until `kdbSet` is called.

In any case, this document is not addressed to API users, so it doesn't matter if it is interesting to API users or not, if it is interesting to the audience of the decision. Which it is, because the only interesting bit here is where `munmap` happens.

  We can still let users access the flag `ELEKTRA_CP_COW`, we just need to clearly document what is forbidden.
  Maybe set the `KEY_FLAG_RO_VALUE` on the original key, so that the API itself detects the error. 
  There is, however, no flag for `keyDel` that we could set.

> ### Data restrictions

Reviewer: I would put this earlier so that it better grouped by topics.

Author: What do you mean by "this"? The "Data restrictions" and "API restrictions" alternatives?

Author: [3dda5d5](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/3dda5d5bfa9b15dc7e48d4218e10f6acede49b12)

> If the keyset is owned by the KDB handle, it should not be as big surprise, if there is extra data in there.
> I certainly wouldn't try to asset anything on the contents of a KeySet that I don't own directly, unless the condition is explicitly documented somewhere.

@markus2330 wrote:

> I disagree, it is actually the same kind of surprise for "More Keys".
> Only the "Fewer Keys" would get fixed.

### Full-blown copy-on-write implementation

Make Elektra's `Key` and `KeySet` data structures copy-on-write.
This requires some major refactoring of code within `libelektra-core`.
Code that does only interact with the data structures via the public `libelektra-core` API should not notice any differences.
The `mmapstorage` plugin will need a major refactoring.

#### Changes to `Key`

For the `Key`, we need to extract everything for the data and name into their own structs.
This is done for memory-management reasons, as we need to track how many keys point to the same data and/or name.

```c
struct _KeyData {
    union {
        char * c;
        void * v;
    } data;

    size_t dataSize;

    uint16_t refs;
};

struct _KeyName {
    char * key;

>     size_t keySize;

Reviewer: Actually we want to get rid of this, see ./2_in_progress/store_name.md

Author: Okay. But what does "in progress" mean in that case? Is somebody actively working on it?

Reviewer: Yes, @mpranj worked on it.

Reviewer 2: AFAIK there has been zero progress on removing the key names. Also "In Progress" is _before_ "Decided" so that's extra confusing and more reason why I suggested to remove "In Progress" in #4639.

>     char * ukey;

Reviewer: I have an additional proposal here: What about having an extra namespace, several (ref-counted) pointers to namespace, dirname and basename. The dirname is what we have as parent key. Then we need to store the parent-key (usually the largest portion) only once, and for all the other keys only the basename.

This way should save plenty of memory (instead of using more). The problem is that we cannot have a `keyName` anymore.

Reviewer 2: Storing the namespace separately could make sense (for different reasons), but that's just a single byte.

Regarding `dirname` & `basename`:
There are many problems with this. The first and biggest is where do we split between `dirname` and `basename`? Finding the optimal split would be quite difficult and would change over time. How would the API for key creation and key name changes look like?

If we really want to go in that direction, I think we actually need to rethink the whole API (again) and see, if we can find a better way to do things.

Author: Nice proposal, but let's try to avoid doing everything at once in this decision.

Reviewer 2: > It is not a single byte. With COW is N*M bytes, N the number of namespaces, M the shared portion of the keyname.

Assuming we remove the escaped key name, the namespace is stored literally as a single byte. There is no point in doing anything other than just copying it.

> I suggested the parent-key above.

But what is _the parent-key_? The term is context-dependent. There is the parent key of a mountpoint, the parent key used in `kdbGet`, etc.

Even if one of those is chosen (parent of mountpoint would be sensible), this still doesn't answer how `keyNew` or `keySetName` would now about this.

> I don't see a connection between "not having keyName (and an iterator instead)" and "need to rethink the whole API".

And that's honestly, why you shouldn't make these decisions. I've dealt with key name changes before. I know it affects _literally everything_. If there is any point to rethink the API that is the point. It is almost impossible to find an API change that would affect more code than key name changes.

And like I explained above, but you may not have realized yet, the changes you proposed **require** API changes. Not just that, but API changes to pretty fundamental functions used all over the code base. This really would be the point to rethink the whole API.

> mpranj can you add these possibilities to the store_name.md decision?

I seriously doubt mpranj wants to be involved in this. AFAIK the task isn't even related to mpranj's previous work. IMO it is a very bad habit of yours to assign such tasks to random people.

Reviewer 2: Since you marked my comment as disruptive, I'll repeat the purely technical and absolutely correct parts again. There is no need to answer it here and now. But if you expect this approach to be added to the decision or to be discussed elsewhere, we need an answer to these questions.

> It is not a single byte. With COW is N*M bytes, N the number of namespaces, M the shared portion of the keyname.

Assuming we remove the escaped key name, the namespace is stored literally as a single byte. There is no point in doing anything other than just copying it.

> I suggested the parent-key above.

But what is _the parent-key_ here? The term is context-dependent. There is the parent key of a mountpoint, the parent key used in kdbGet, etc.

Even if one of those is chosen (parent of mountpoint would be sensible), this still doesn't answer how keyNew or keySetName would now about this. I don't see how this would work without API changes to very basic APIs used all over the code base.

    size_t keyUSize;

    uint16_t refs;
};

struct _Key {
    struct _KeyData * keyData;    
    struct _KeyName * keyName;
    KeySet * meta;
    keyflag_t flags;

    uint16_t refs;
};
```

The following calculations are based on the AMD64 platform:

In the data structures above, an empty key does have
- with `_keyData` and `_keyName` allocated: 96 bytes.

> - with `_keyData` and `_keyName` `NULL`: 32 bytes.

Reviewer: :+1: please also write another common case: keyName allocated but no _keyData

Author: [bbc8302](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/bbc83023b7dd2d5b31684eb01416ec4d61a06852)

`(sizeof(_Key) + name + 1 + ukey + value + 1) * 2` =
(64 + 28 + 1 + 25 + 34 + 1) * 2 = 306 bytes.

#### Changes to `KeySet`

For `KeySet`, we need to split out everything to do with the stored keys into a separate datastructure.
This includes the array itself, the sizes and the hashmap.

Why don't we just add the number of references to the original `KeySet`?

- If we delete a copied KeySet, we don't know which KeySet is the original, so we couldn't decrement the counter. 
  This could be dealt with storing a pointer to the original KeySet.
- If the original KeySet is deleted, we don't know which other KeySets point at the data, so updating their count would not work
- In similar fashion, if you update the original KeySet, the copied KeySets will also contain the new data (if the memory address does not change). 
  This is unexpected behaviour.

```c
struct _KeySetData {
    struct _Key ** array;
    size_t size;  /**< Number of keys contained in the KeySet */
    size_t alloc; /**< Allocated size of array */

    Opmphm * opmphm;
    OpmphmPredictor * opmphmPredictor;

    uint16_t refs; /**< Reference counter */
};

struct _KeySet {
    struct _KeySetData * data;

>     struct _Key * cursor; /**< Internal cursor */
>     size_t current;		  /**< Current position of cursor */

Reviewer: As @kodebach already wrote this should be removed.

```suggestion
```

Author: [c81d549](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/c81d5493f6c3b5a5962f08fa2bb1b4bc1714eefb)

	
    ksflag_t flags;

    uint16_t refs; /**< Reference counter */
};
```

The following calculations are based on the AMD64 platform:

An empty keyset with the datastructure above has 80 bytes.
An empty keyset with the current implementation has 64 bytes.

A copied, non-modified keyset with the datastructure above has always 32 bytes.
A copied, non-modified keyset with the current implementation has at least 64 bytes (for an empty keyset).

> 

Reviewer: Sizes are not the only criteria, a disadvantage of this approach is the probably higher number of allocations.

Reviewer 2: > higher number of allocations.

That is an issue I was concerned with too. But do we actually have any data to show that this is a problem? Is it really the high number of individual allocations or is it actually the larger number of indirections?

If the problem is just a large number of small allocations, then we could maybe implement a pool allocator. Then there would just be one (or multiple) bigger pools for each type of struct.

Author: What exactly are our concerns with the number of allocations? @markus2330 can you formulate a research question for this? I can then do some research and benchmarking to find the answer for it.

Reviewer: > What exactly are our concerns with the number of allocations? @markus2330 can you formulate a research question for this?

The concerns are basically covered with CPU+memory overhead. I was throwing this in as it is easy to calculate during the decision phase. Many more allocations are definitely not favorable for CPU+memory overhead and give a more complete picture than simply looking at bytes.

Ideally you could make a table with the different COW options and their estimated resource usage (bytes+nr of allocations). For the thesis this will be nice; if it will already help for the decision, I have to admit that I don't know.

### Fewer Keys

When doing a second `kdbGet` with a new keyset no keys will be returned when no backends report changed data, because kdb internally thinks the data is already up-to-date.
A unit test by @atmaxinger:

```c
static void test_doubleGet (void)
{
	printf("running %s\n", __func__);

	// Setup
	Key * parentKey = keyNew("/somewhere", KS_END);
	KeySet * ks = ksNew(0, KS_END);
	KDB * kdb = kdbOpen (ksNew(0, KS_END), parentKey);
	kdbGet (kdb, ks, parentKey);
	ksAppendKey (ks, keyNew ("user:/somewhere", KEY_VALUE, "abc", KEY_END));
	ksAppendKey (ks, keyNew ("user:/somewhere/key", KEY_VALUE, "xyz", KEY_END));
	kdbSet (kdb, ks, parentKey);
	kdbClose (kdb, parentKey);

	// Scenario
	kdb = kdbOpen (ksNew(0, KS_END), parentKey);

	KeySet * ks1 = ksNew (0, KS_END);
	KeySet * ks2 = ksNew (0, KS_END);

	kdbGet (kdb, ks1, keyNew("/somewhere", KEY_END));
	succeed_if (ksLookupByName (ks1, "/somewhere/key", 0) != NULL, "should find key (1)");
	kdbGet (kdb, ks2, keyNew("/somewhere", KEY_END));
	succeed_if (ksLookupByName (ks2, "/somewhere/key", 0) != NULL, "should find key (2)");
}
```

It was found unexpected that the second assert - should find key (2) - will fail.

## Constraints

> - memory consumption must be low for `kdbGet`, see [4. Goal: Performance](/doc/GOALS.md), in particular, deep duplication is too expensive

Reviewer 2: > For non-optional parts of Elektra, also non-POSIX systems must be supported

We should clarify what "non-POSIX" means. AFAIK we only support POSIX and MinGW (subset of POSIX). At least that is what we test for. Last I checked, even `libelektra-core` did not compile on a pure Windows system, so unless there is something I missed, I don't think we can claim Elektra runs on "anything with libc".

A `keyCopy()` without `KEY_CP_COW` from an COW key will create a deep copy of the key.
These keys are "normal" non-COW keys and can live on their own.

Every `key*()` method that modifies data on a COW-copied key will need to allocate new memory for this data and remove the `KEY_FLAG_COW_DATA` flag.
Every `key*()` method that modifies the name of a COW-copied key will need to allocate new memory for this name and remove the `KEY_FLAG_COW_NAME` flag.
Every `key*()` method that modifies metadata needs to make sure that the same happens for metakeys.

Keysets are not copy-on-write.
A `ksDeepDup()` of a keyset with COW keys will create a keyset with deep-copied non-COW keys.
Internally we may need a `ksCowDup()` function to create a keyset with copy-on-write keys from another keyset.
Whether this function will be part of the public API is a point for discussion.

### Full-blown copy-on-write implementation

Make Elektra's `Key` and `KeySet` data structures copy-on-write.
This requires some major refactoring of code within `libelektra-core`.
Code that does only interact with the data structures via the public `libelektra-core` API should not notice any differences.

> The `mmapstorage` plugin will need a major refactoring.

Reviewer 2: > The `mmapstorage` already needs to be fixed.

Why? What exactly is broken? AFAIK it works perfectly fine, [even as the default storage plugin](https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/master/858/pipeline/1062).

The `cache` plugin on the other hand, is broken and needs to be fixed because of the `new-backend` changes. Maybe that's what you're thinking of.

> **Note** `cache` and `mmapstorage` are separate plugins. `cache` depends on `mmapstorage`, but `mmapstorage` works without `cache`. Using the terms interchangeably is confusing and wrong.

>
> This approach wouldn't be very performant (since it uses disk IO), but especially if we can use the cache data, it should be pretty easy to do.

### MMAP Cache without parent key

We make the mmap cache non-optional and only use a single cache, caching everything.
We remove the parent key of `kdbGet` and `kdbSet` and always return the keyset of the whole KDB.

### Data restrictions

@kodebach wrote:

> Make all the keys returned by kdbGet completely read-only.
> To change the data you need to append an entirely new key to replace the existing one.
> Then we just need to keep a shallow copy internally.

### API restrictions

@kodebach wrote:

> Change the API and remove KeySet from kdbGet and kdbSet also option 4 in [the operation sequences decision](../0_drafts/operation_sequences.md).
> If the keyset is owned by the KDB handle, it should not be as big surprise, if there is extra data in there.

> > I certainly wouldn't try to asset anything on the contents of a KeySet that I don't own directly, unless the condition is explicitly documented somewhere.

Reviewer 2: ```suggestion
> I certainly wouldn't try to assert anything on the contents of a KeySet that I don't own directly, unless the condition is explicitly documented somewhere.
```

### Full-blown copy-on-write implementation

Make Elektra's `Key` and `KeySet` data structures copy-on-write.
This requires some major refactoring of code within `libelektra-core`.
Code that does only interact with the data structures via the public `libelektra-core` API should not notice any differences.
The `mmapstorage` plugin needs to be fixed.

Unlike "In-Memory COW Cache" keyDup, keyCopy, ksCopy and ksDup will always use COW.
`ksCopy` and `ksDup` is needed for (de)duplication of metadata.
Furthermore, the API has better usability if Key and KeySet behave the same, especially for bindings where duplication might happen implicitly.

#### Changes to `Key`

For the `Key`, we need to extract everything for the data and name into their own structs.
This is done for memory-management reasons, as we need to track how many keys point to the same data and/or name.

```c
struct _KeyData {
    union {
        char * c;
        void * v;
    } data;

    size_t dataSize;

    uint16_t refs;
};

struct _KeyName {
    char * key;
    size_t keySize;

    char * ukey;
    size_t keyUSize;

    uint16_t refs;
};

struct _Key {
    struct _KeyData * keyData;

>     struct _KeyName * keyName;

Reviewer 3: Some thoughts regarding moving name and data to separate structures:

1. If they are a separate entity, `mmapstorage` will need a flag once again for each of those. This is used to mark whether the data is in an mmap region or not. (or we find some bit somewhere that we can steal for this purpose)
2. Adding more indirections is probably not going to help performance. (I understand that we save memory here)

Author: I've added this to the discussion

32\*2 + 40 + 24 + 28 + 1 + 25 + 34 + 1 = 217 bytes.

For a single key, it would take the current implementation:

`sizeof(_Key) + name + 1 + ukey + value + 1` =
64 + 28 + 1 + 25 + 34 + 1 = 153 bytes

For two keys, the current implementation would take:

`(sizeof(_Key) + name + 1 + ukey + value + 1) * 2` =
(64 + 28 + 1 + 25 + 34 + 1) \* 2 = 306 bytes.

#### Changes to `KeySet`

For `KeySet`, we need to split out everything to do with the stored keys into a separate datastructure.
This includes the array itself, the sizes and the hashmap.

Why don't we just add the number of references to the original `KeySet`?

- If we delete a copied KeySet, we don't know which KeySet is the original, so we couldn't decrement the counter.
  This could be dealt with storing a pointer to the original KeySet.
- If the original KeySet is deleted, we don't know which other KeySets point at the data, so updating their count would not work
- In similar fashion, if you update the original KeySet, the copied KeySets will also contain the new data (if the memory address does not change).
  This is unexpected behaviour.

```c
struct _KeySetData {

>     struct _Key ** array;

Reviewer 3: While we're making these fundamental changes...

One thing I was curious about, but I thought these kind of changes were off the table:

```c
struct _KeySetData {
    struct _KeyName ** array;
    struct _KeyData ** array;

    size_t keySize;
    size_t keyDataSize;

    [...]
}
```

In that case we could do the `mmap` pointer update using vector extensions and probably speed it up significantly.

Reviewer 2: Yeah, I've also though about similarly radical changes in the past, but I've never thought we could actually agree to do something as big as that.

Reviewer: Big changes that'll be done in the near future by you before 1.0 are totally possible.

As performance is the lowest of our 4 goals, and memory usage was in the last benchmark a **much** bigger problem than CPU, I don't think a decision that has a focus to speed without a clear usability benefit, has any chance.

If the change is an improvement for e.g. the API user and the speed-up is only a side effect, I cannot wait for your proposal in form of a decision PR.

Reviewer 2: > Big changes that'll be done in the near future by you before 1.0 are totally possible.

Not the ones I've thought of... Even if I had the time to implement them. Basically, I've again and again come to the conclusion that our biggest limitation is the separate `Key` structure (or at least the fact that it is public API). More specifically, that `Key` is actually a whole (key,value) pair despite its name.

For example a [Radix tree](https://en.wikipedia.org/wiki/Radix_tree) (or [HAT-trie](https://en.wikipedia.org/wiki/HAT-trie)) would be prefect for `KeySet`, but it just doesn't work. You'd effectively store the name twice, because you still need the full name in `Key`.

> I think this proposal might break COW for keys.

The much bigger problem here (as alluded to above) is that there is no `Key *` in the sturcture. This immediately makes this snippet impossible

```c
Key * k = keyNew ("user:/foo", KEY_END);
ksAppendKey (ks, k);
assert (k == ksLookupByName (ks, "/foo", 0));
```

More generally: How would all the `ks*` functions that return a `Key *` work at all?
You'd have to allocate a new `Key *` or take a `Key * dest` as argument or use some fixed `Key *` that is only valid until the next call. All of which would be API changes so big and complex that you'd almost have to rewrite the whole codebase and definitely not something that could be automated.

    copy over the non-cut keys into `dest->data` from the old `dest->data`
  ```
- `ksPop(dest)`:

  ```
  if `dest->data->refs` == 0 (meaning no other keysets points to it):
    pop the last key directly on the keys
  else if `dest->data->refs` > 0: (there are other keysets sharing the same data):
    decrease `dest->data->refs` by 1
    create a new `dest->data`
    copy over all but the last keys from the old `dest->data`
  ```

- `ksLookup()`:
  - if the `DEL` or `POP` flag is specified, do the COW-stuff as described multiple times now in the operations above

#### Optimizations

This approach requires more allocation than previously.
We have not benchmarked whether this is big of an issue.
One optimization could be an expanding "pool" of `_KeySetData`, `_KeyData` and `_KeyName`.
We could then allocate multiple of them at the same time, and borrow and give back instance from and to the pool.

## Memory comparison of COW approaches

The following calculations are based on the AMD64 platform.
All results are in bytes unless stated otherwise.

Example key: `user:/hosts/ipv6/example.com = 2606:2800:220:1:248:1893:25c8:1946`

We want to measure the following properties:

> - Empty KeySet: size of a simple malloc of the keyset struct

Reviewer 2: Either remove the KeySet column, or add a second table similar to the one for Keys that fully shows the situation for KeySets, i.e., "Empty KeySet", "Empty KeySet + extra structs", "KeySet with 1 key", etc.

The current table suggests the "full-blown COW" needs less memory for a keyset, when this is not the case. It's only the case, when the keyset shares it's keys with another.

Author: Agree, I added another table for the keyset.

- Elektra doesn't require MMAP

**Cons:**

- Lifetime of a copied COW key MUST be less than the key it was copied from.
  We can not track how many keys point to the same data this way, so we can only free data if the key does not have the COW flag.
  If the original key gets deleted, using a COW key that points to the same data will lead to corrupt data.
  The same is true for updating values of the original key.
  
  This is only problematic if we want to use COW for keys outside of `KDB`.

>   If it is only for use within `KDB`, especially for usage as internal cache and in change tracking, we always know that the original keys are going to last as long as the `KDB` instance.

Reviewer 2: ```suggestion
  If it is only for use within `KDB`, especially for usage as internal cache and in change tracking, we could always guarantee that the original keys are going to last as long as the `KDB` instance.
```

It is not automatically guaranteed that keys last until `kdbClose`. For change tracking it would probably make much more sense if the keys were only valid until the next `kdbGet`/`kdbSet`. Otherwise we would have to keep all versions of keys internally until `kdbClose`.

Also I just checked and it turns out that data returned by `mmapstorage` is only valid until the next `kdbSet`, because that's when `munmap` is called. But I think (at least for the cache) we could move that to `close`.

Sorry for the confusion because of my earlier comments, should have checked the code first.

  If they want to continue using them afterwards, they'd have to deep copy them.
  
  Triggering the delete problem:
  ```c
  Key * originalKey;
  Key * copiedKey;
  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
  
  assert (keyString(copiedKey) == keyString(originalKey));
  keyDel (originalKey);
  
  keyString(copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to freed memory
  ```
 
  Triggering the update problem:
  ```c
  Key * originalKey;
  Key * copiedKey;
  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
  
  assert (keyString(copiedKey) == keyString(originalKey));
  keySetString (originalKey, "new value!");
  
  keyString(copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to potentially freed memory
  ```

  The same problems in principle exist for `mmapstorage`, however this is just a storage plugin and the user does not have any API to create COW keys themselves.
  We can still let users access the flag `ELEKTRA_CP_COW`, we just need to clearly document what is forbidden.

>   Maybe set the `KEY_FLAG_RO_VALUE` on the original key, so that the API itself detects the error. 

Author: Yes, but we will clearly state that in the documentation.

@kodebach wrote:

> Change the API and remove KeySet from kdbGet and kdbSet also option 4 in [the operation sequences decision](../0_drafts/operation_sequences.md).
> If the keyset is owned by the KDB handle, it should not be as big surprise, if there is extra data in there.
> I certainly wouldn't try to assert anything on the contents of a KeySet that I don't own directly, unless the condition is explicitly documented somewhere.

@markus2330 wrote:

> I disagree, it is actually the same kind of surprise for "More Keys".
> Only the "Fewer Keys" would get fixed.

### Copy On Write

We keep a duplicated keyset in-memory and tag the keys as copy-on-write (COW).
From this keyset, we use `ksBelow` to return the correct keyset.
If the user tries to change the value or metadata of these keys, the data gets duplicated.
I.e. the original keyset is not changed.
The name is not relevant.
It is always read-only, because the key is in at least one keyset (the internal one).

> Possible copy-on-write implementations are described in [another discussion](../1_in_discussion/copy_on_write.md).

Reviewer 2: ```suggestion
Possible copy-on-write implementations are described in [another decision](../1_in_discussion/copy_on_write.md).
```

> I disagree, it is actually the same kind of surprise for "More Keys".
> Only the "Fewer Keys" would get fixed.

### Full-blown copy-on-write implementation

Make Elektras `Key` and `KeySet` datastructures copy-on-write.
This requires some major refactoring of code within `libelektra-core`.
Code that does only interact with the datastructures via the public `libelektra-core` API should not notice any differences.
The `mmapstorage` plugin will need a major refactoring.

### Changes to `Key`

For the `Key`, we need to extract everything for the data and name into their own structs.
This is done for memory-management reasons, as we need to track how many keys point to the same data and/or name.

In the data structures below, an empty key does have 96 bytes.
An empty key of the current implementation has 64 bytes.

A copied, non-modified key with the data structure below does always have 32 bytes.

> A copied, non-modified key of the current implementation has at least 64 bytes (for an empty key), but in reality much more as the name and the data are also copied.

Author: [512a805](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/512a805f77aef6db20676cd56a0eed600b76115e)

```c
struct _KeyData {
    union {
        char * c;
        void * v;
    } data;

    size_t dataSize;

    uint16_t refs;
    uint16_t reserved;
};

struct _KeyName {
    char * key;
    size_t keySize;

    char * ukey;
    size_t keyUSize;

    uint16_t refs;
    uint16_t reserved;
};

struct _Key {
    struct _KeyData * keyData;    
    struct _KeyName * keyName;
    KeySet * meta;
    keyflag_t flags;

    uint16_t refs;
    uint16_t reserved;
};
```

### Changes to `KeySet`

> For `KeySet`, we need to split out everything to do with the stored keys into a separate datastructure.

Author: Two things,

1. If we delete a copied KeySet, we don't known which KeySet is the original, so we couldn't decrement the counter. This could be dealt with storing a pointer to the original KeySet.

2. If the original KeySet is deleted, we don't know which other KeySets point at the data, so updating their count would not work

3. In similar fashion, if you update the original KeySet, the copied KeySets will also contain the new data (if the memory address does not change). This is unexpected behaviour.

> For `KeySet`, we need to split out everything to do with the stored keys into a separate datastructure.

Author: [9c035ac](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/9c035ac17415843d155e1cca7159b1ff39dedc93)

This includes the array itself, the sizes and the hashmap.

An empty keyset with the datastructure below has 80 bytes.
An empty keyset with the current implementation has 64 bytes.

A copied, non-modified keyset with the datastructure below has always 32 bytes.
A copied, non-modified keyset with the current implementation has at least 64 bytes (for an empty keyset).

```c
struct _KeySetData {
    struct _Key ** array;
    size_t size;  /**< Number of keys contained in the KeySet */
    size_t alloc; /**< Allocated size of array */

    Opmphm * opmphm;
    OpmphmPredictor * opmphmPredictor;

    uint16_t refs; /**< Reference counter */
    uint16_t reserved; /**< Reserved for future use */
};

struct _KeySet {
    struct _KeySetData * data;

    struct _Key * cursor; /**< Internal cursor */

>     size_t current;		  /**< Current position of cursor */

Author: [c81d549](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/c81d5493f6c3b5a5962f08fa2bb1b4bc1714eefb)
